NeoOnnx/src/Optimizer.h

```diff
@@ -0,0 +1,75 @@
+/* Copyright © 2017-2020 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include "common.h"
+#include <NeoOnnx/NeoOnnxImport.h>
+
+namespace NeoOnnx {
+
+class IOptimizer : public IObject {
+public:
+	explicit IOptimizer( CDnn& graph ) : graph( graph ) {}
+	IOptimizer( IOptimizer&& ) = delete;
+	IOptimizer( const IOptimizer& ) = delete;
+
+	virtual void Apply() = 0;
+
+protected:
+	CDnn& graph;
+	CArray<CPtr<CBaseLayer>> layersToRemove{};
```

Comment:
You give direct access to this field, but at the same time non-trivial methods for adding layers to it.
I think it's better to hide it in `private` and add emthods like `Clear == layersToRemove.DeleteAll();` and `ApplyRemoval` which deletes layers from `graph`.

_minor_ The appearance of layer in this array doesn't _guarantee_ that it will be deleted (you're detecting a block of 10 layers, and when you've found 3 of them you add them and only after you failed to detect 4'th layer you will clean the array. That's why it's better to name more neutral, like `markedLayers` `selectedLayers`, and (as I mentioned before) name all the methods like `ClearSelection` `RemoveSelectedLayers`, flag `justCheck` rename to `addToSelection` or simply `mark/select`

Reply:
fixed